File: Easy/Python/169. Majority Element.py
'''
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.
'''

class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        vote = 1
        length = len(nums)
        contestent = nums[0]

        for ind in range(1,length):
            if nums[ind] == contestent:
                vote += 1
            elif vote < 0:
                contestent = nums[ind]
                vote = 1
            else:
                vote -= 1

        return contestent

File: Easy/Javascript/169. Majority Element.js
/**Given an array nums of size n, return the majority element.
 * The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array. */

/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function (nums) {
  let count = 1;
  let candidate = nums[0];

  for (let ind = 1; ind < nums.length; ind++) {
    if (count === 0) {
      candidate = nums[ind];
      count = 1;
    } else if (nums[ind] === candidate) {
      count += 1;
    } else {
      count -= 1;
    }
  }

  return candidate;
};

console.log(majorityElement([2, 2, 1, 1, 1, 2, 2]));
